activerecord/lib/active_record/connection_adapters/metadata_registry.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+require "digest/md5"
+
+module ActiveRecord
+  module ConnectionAdapters # :nodoc:
+    module MetadataRegistry
+      REGISTRY = {}
+
+      class << self
+        def fetch(type, *args)
+          key = Digest::MD5.hexdigest(Marshal.dump([type.name, args]))
+          REGISTRY[key] ||= yield.freeze
```

Comment:
Why do we need dump and hexdigest? Why won't just `REGISTRY[[ type.name, args ]]` suffice?

Reply:
Yeah good point. Actually now that I dug a bit more in these classes, I saw that they all define equality pretty well, so I'm tempted to simply let them be instanciated, and then dedup after.